File: BCL/Common/EnumExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Reflection
Imports System.Linq


Namespace Common


    <AttributeUsage(AttributeTargets.Enum Or AttributeTargets.Field)> _
    Public Class EnumDisplayName
        Inherits Attribute
        Dim m_DisplayName As String = String.Empty
        Public Sub New(ByVal DisplayName As String)
            m_DisplayName = DisplayName
        End Sub
        Public Overrides Function ToString() As String
            Return m_DisplayName
        End Function
    End Class

    <AttributeUsage(AttributeTargets.Enum Or AttributeTargets.Field)> _
    Public Class EnumDBValue
        Inherits Attribute
        Dim m_DBValue As String = String.Empty
        Public Sub New(ByVal DBValue As String)
            m_DBValue = DBValue
        End Sub
        Public Overrides Function ToString() As String
            Return m_DBValue
        End Function
    End Class

    Public Class EnumItem
        Public Property Name As String
        Public Property Value As Integer
        Public Overrides Function ToString() As String
            Return Name
        End Function
    End Class

    Public Module EnumExtensions

        'Return the Enum entry for the specified string
        'TryParse is case-sensitive, so use LINQ instead
        Public Function Entry(Of T)(ByVal dbvalue As String) As T
            Dim fis As MemberInfo() = GetType(T).GetFields()

            For Each fi As FieldInfo In fis
                Dim attr As Attribute = Attribute.GetCustomAttribute(fi, GetType(EnumDBValue))
                If Not attr Is Nothing And TypeOf attr Is EnumDBValue Then
                    If DirectCast(attr, EnumDBValue).ToString = dbvalue Then
                        Return DirectCast([Enum].Parse(GetType(T), fi.Name), T)
                    End If
                End If
            Next

            Throw New Exception("Not found")
        End Function


        Public Function StringAttribute(ByVal Value As [Enum], ByVal AttributeName As String) As String
            Dim Result = String.Empty
            Dim Field As FieldInfo = Value.GetType.GetField(Value.ToString, BindingFlags.GetField Or BindingFlags.Public Or BindingFlags.Static)
            '' Dim attribs = Field.GetCustomAttributes(True)
            Dim dn = (From item In Field.GetCustomAttributes(True) Where item.GetType().Name = AttributeName Select item).FirstOrDefault
            If (Not IsNothing(dn)) Then Result = dn.ToString
            Return Result
        End Function

        <Extension()> _
        Public Function DisplayName(ByVal Value As [Enum]) As String
            Dim Result = StringAttribute(Value, "EnumDisplayName")
            Return Result
        End Function

        <Extension()> _
        Public Function DBVAlue(ByVal Value As [Enum]) As String
            Dim Result = StringAttribute(Value, "EnumDBValue")
            Return Result
        End Function


        'Move to separate function to avoid issue with late binding when Option Strict On
        Public Function GetNames(s As [Enum]) As String()
            Return CType([Enum].GetNames(s.GetType), String())
        End Function

        ''Move to separate function to avoid issue with late binding when Option Strict On

        <Extension()> _
        Public Function ItemList(ByVal s As [Enum], ByVal IncludeUnknown As Boolean, ByVal UseDisplayName As Boolean) As List(Of EnumItem)
            Dim Result As New List(Of EnumItem)
            Dim names = GetNames(s)
            Dim values = [Enum].GetValues(s.GetType)
            For i As Integer = 0 To names.Length - 1
                If (IncludeUnknown OrElse names(i).ToLower <> "unknown") Then
                    If (UseDisplayName) Then
                        Dim itemName = DisplayName(CType(values(i), [Enum]))
                        Result.Add(New EnumItem With {.Name = itemName, .Value = values(i)})
                    Else
                        Result.Add(New EnumItem With {.Name = names(i), .Value = values(i)})
                    End If
                End If
            Next
            Return Result
        End Function
    End Module

End Namespace

File: EnumLib/EnumExtensions.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Linq;
using System.Text;
using System.Reflection;

namespace EnumLib
{
	public class EnumItem
	{
		public string Name { get; set; }
		public int Value { get; set; }
		public override string ToString()
		{
			return Name;
		}
	}

	public static class EnumExtensions
	{
		//Return the Enum entry for the specified string
		//TryParse is case-sensitive, so use LINQ instead
		public static T Entry<T>(string dbvalue)
		{
			MemberInfo[] fis = typeof(T).GetFields();

			foreach (FieldInfo fi in fis)
			{
				Attribute attr = Attribute.GetCustomAttribute(fi, typeof(EnumDBValue));
				if ((attr != null) & attr is EnumDBValue)
				{
					if ( (attr as EnumDBValue).ToString() == dbvalue)
					{
						return (T)Enum.Parse(typeof(T), fi.Name);
					}
				}
			}

			throw new Exception("Not found");
		}

		public static string StringAttribute(Enum value, string attributeName)
		{
			dynamic Result = string.Empty;
			FieldInfo Field = value.GetType().GetField(value.ToString(), BindingFlags.GetField | BindingFlags.Public | BindingFlags.Static);
			//' Dim attribs = Field.GetCustomAttributes(True)
			dynamic dn = (from item in Field.GetCustomAttributes(true)
						  where item.GetType().Name == attributeName 
						  select item).FirstOrDefault();
			if ((dn != null))
				Result = dn.ToString;

			return Result;
		}


		public static string DisplayName(this Enum value)
		{
			dynamic Result = StringAttribute(value, "EnumDisplayName");
			return Result;
		}


		public static string DBVAlue(this Enum value)
		{
			dynamic Result = StringAttribute(value, "EnumDBValue");
			return Result;
		}

		//Move to separate function to avoid issue with late binding when Option Strict On

		//'Move to separate function to avoid issue with late binding when Option Strict On

		public static List<EnumItem> ItemList(this Enum s, bool includeUnknown = false , bool useDisplayName = false)
		{
			List<EnumItem> Result = new List<EnumItem>();
			dynamic names = Enum.GetNames(s.GetType());
			dynamic values = Enum.GetValues(s.GetType());
			for (int i = 0; i <= names.Length - 1; i++)
			{
				if ((includeUnknown || names(i).ToLower != "unknown"))
				{
					if ((useDisplayName))
					{
						dynamic itemName = DisplayName((Enum)values(i));
						Result.Add(new EnumItem
						{
							Name = itemName,
							Value = values(i)
						});
					}
					else
					{
						Result.Add(new EnumItem
						{
							Name = names(i),
							Value = values(i)
						});
					}
				}
			}
			return Result;
		}
	}
}
